File: zad2/zadanie2.kt
import java.util.*

fun main() {
    print("Введите N: ")
    val line = readLine()
    val n = line?.toIntOrNull()
    if (n == null || n <= 0) {
        println("Не найдено")
        return
    }
    val zayvki = zapolTime(n)
    SortZayavki(zayvki)
    val nomera = CountMaxZayavki(zayvki)
    printNomera(nomera)
}

fun lessZayavka(firstZ: IntArray, secondZ: IntArray): Boolean {
    if (firstZ[0] != secondZ[0]){
        return firstZ[0] < secondZ[0]
    } 
    if (firstZ[1] != secondZ[1]){
        return firstZ[1] < secondZ[1]
    } 
    return firstZ[2] < secondZ[2]
}
fun zapolTime(n: Int): MutableList<IntArray> {
    val zayvki = mutableListOf<IntArray>()
    for (i in 1..n) {
        val line = readLine()?.trim() ?: continue
        if (line.isEmpty()) continue
        val parts = line.split(Regex("\\s+")).map { it.toInt() }
        if (parts.size != 2) continue
        val start = parts[0]
        val end = parts[1]
        zayvki.add(intArrayOf(end, start, i))
    }
    return zayvki
}
fun SortZayavki(zayvki: MutableList<IntArray>) {
    zayvki.sortWith(compareBy({ it[0] }, { it[1] }, { it[2] }))
}

fun CountMaxZayavki(zayvki: List<IntArray>): List<Int> {
    val nomera = mutableListOf<Int>()
    var lastEnd = -1
    for (z in zayvki) {
        val end = z[0]
        val start = z[1]
        val index = z[2]
        if (start >= lastEnd) {
            nomera.add(index)
            lastEnd = end
        }
    }
    nomera.sort()
    return nomera
}

fun printNomera(nomera: List<Int>) {
    print("Результат: ${nomera.size}, ")
    for (x in nomera) print("$x ")
    println()
}

File: zad2/zadanie2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void zapolTime(int n, vector<pair <int,int>>& time);
void printNomera(const vector<int>& nomera);
int CountElements(const vector<int>& v);
vector<pair<int, pair<int,int>>> SortZayavki(const vector<pair<int, int>>& time);
vector<int> CountMaxZayavki(const vector<pair<int,pair<int,int>>>& zayvki);

int main(){
    vector<pair<int,int>> time;
    vector<int> nomera;
    int n {};
    cout << "Введите N\n";
    cin >> n;
    cout << "\nВведите время начало и конца\n";
    zapolTime(n, time);
    cout << "\n Максимальное количество встреч" << endl;
    vector<pair<int, pair<int,int>>>zayvki = SortZayavki(time);
    nomera = CountMaxZayavki(zayvki);
    cout << "Результат: " << CountElements(nomera) << ",  ";
    printNomera(nomera);
}


int CountElements(const vector<int>& nomera) {
    int count {};
    for (int val : nomera){
        count++;
    } 
    return count;
}

vector<pair<int, pair<int,int>>> SortZayavki(const vector<pair<int, int>>& time){
    vector<pair<int, pair<int,int>>> zayvka;
    for (int i=0; i<time.size();++i){
        int start = time[i].first;
        int end = time[i].second;
        zayvka.push_back({end, {start,i+1}});
    }
    sort(zayvka.begin(),zayvka.end());
    return zayvka;
}

vector<int> CountMaxZayavki(const vector<pair<int,pair<int,int>>>& zayvki){
    vector <int> chose;
    int lEnd{-1};
    for (pair<int, pair<int,int>> val : zayvki){
        int end = val.first;
        int start = val.second.first;
        int index = val.second.second;
        if (start>=lEnd){
            chose.push_back(index);
            lEnd = end;
        }
    }
    sort(chose.begin(),chose.end());
    return chose;  
}

void zapolTime( int n, vector<pair<int,int>>& time){
    int startTime{};
    int endTime{};
    for (int i=0; i< n; ++i){
        cin>>startTime;
        cin>>endTime;
        time.push_back({startTime, endTime});
    }
}

void printNomera(const vector<int>& nomera){
    for (int val : nomera){
        cout << val << " ";
    }
}

